Use SpannerConnectionMetadata in the connection

Connection should not return null. It should instead return `SpannerConnectionMetadata`, a trivial implementation that does not change with connection, implemented in #163.

The task is:
1) Define a singleton INSTANCE in `SpannerConnectionMetadata`, similar to the one in [H2 implementation](https://github.com/r2dbc/r2dbc-h2/commit/d6a0bdbd7ddd9ae437c5538c8ebddfe472aaeec2).
2) Return the static instance from `SpannerConnection.getMetadata()`:
https://github.com/GoogleCloudPlatform/cloud-spanner-r2dbc/blob/572212ca74d827cdace1a6af91e0ccce42d80fba/cloud-spanner-r2dbc/src/main/java/com/google/cloud/spanner/r2dbc/SpannerConnection.java#L264
3) Write tests for both changes.

@elefeint I am facing issue while building this project with maven. Please let me know how can I make the build successful. I am using my service account for building this project with project owner role. Whatever error I am facing sharing as follows.

[ERROR] com.google.cloud.spanner.r2dbc.it.SpannerDdlIT.testCreateAndDrop  Time elapsed: 1.295 s  <<< ERROR!
io.r2dbc.spi.R2dbcNonTransientResourceException: NOT_FOUND: Instance not found: projects/clef-team-project/instances/reactivetest
Caused by: io.grpc.StatusRuntimeException: NOT_FOUND: Instance not found: projects/clef-team-project/instances/reactivetest

[ERROR] com.google.cloud.spanner.r2dbc.it.SpannerDdlIT.testDdlErrorPropagation  Time elapsed: 1.01 s  <<< ERROR!
io.r2dbc.spi.R2dbcNonTransientResourceException: NOT_FOUND: Instance not found: projects/clef-team-project/instances/reactivetest
Caused by: io.grpc.StatusRuntimeException: NOT_FOUND: Instance not found: projects/clef-team-project/instances/reactivetest

Please suggest me how to resolve this issue.

--- a/cloud-spanner-r2dbc/src/main/java/com/google/cloud/spanner/r2dbc/SpannerConnection.java
+++ b/cloud-spanner-r2dbc/src/main/java/com/google/cloud/spanner/r2dbc/SpannerConnection.java
@@ -24,7 +24,6 @@ import com.google.spanner.v1.TransactionOptions;
 import com.google.spanner.v1.TransactionOptions.ReadWrite;
 import io.r2dbc.spi.Batch;
 import io.r2dbc.spi.Connection;
-import io.r2dbc.spi.ConnectionMetadata;
 import io.r2dbc.spi.IsolationLevel;
 import io.r2dbc.spi.ValidationDepth;
 import java.util.concurrent.atomic.AtomicLong;
@@ -260,8 +259,7 @@ public class SpannerConnection implements Connection, StatementExecutionContext 
   }
 
   @Override
-  public ConnectionMetadata getMetadata() {
-    // TODO: https://github.com/GoogleCloudPlatform/cloud-spanner-r2dbc/issues/163
-    return null;
+  public SpannerConnectionMetadata getMetadata() {
+    return SpannerConnectionMetadata.INSTANCE;
   }
 }

--- a/cloud-spanner-r2dbc/src/main/java/com/google/cloud/spanner/r2dbc/SpannerConnectionMetadata.java
+++ b/cloud-spanner-r2dbc/src/main/java/com/google/cloud/spanner/r2dbc/SpannerConnectionMetadata.java
@@ -24,7 +24,9 @@ import io.r2dbc.spi.ConnectionMetadata;
  * <p>Effectively the same as {@link SpannerConnectionFactoryMetadata}, as there is no
  * connection-level difference in database behavior.
  */
-public class SpannerConnectionMetadata implements ConnectionMetadata {
+public enum SpannerConnectionMetadata implements ConnectionMetadata {
+
+  INSTANCE;
 
   @Override
   public String getDatabaseProductName() {

--- a/cloud-spanner-r2dbc/src/test/java/com/google/cloud/spanner/r2dbc/SpannerConnectionMetadataTest.java
+++ b/cloud-spanner-r2dbc/src/test/java/com/google/cloud/spanner/r2dbc/SpannerConnectionMetadataTest.java
@@ -27,11 +27,12 @@ public class SpannerConnectionMetadataTest {
 
   @Test
   public void productNameIsCorrect() {
-    assertThat(new SpannerConnectionMetadata().getDatabaseProductName()).isEqualTo("Cloud Spanner");
+    SpannerConnectionMetadata spannerConnectionMetadata = SpannerConnectionMetadata.INSTANCE;
+    assertThat(spannerConnectionMetadata.getDatabaseProductName()).isEqualTo("Cloud Spanner");
   }
 
   @Test
   public void productVersionIrrelevant() {
-    assertThat(new SpannerConnectionMetadata().getDatabaseVersion()).isEqualTo("n/a");
+    assertThat(SpannerConnectionMetadata.INSTANCE.getDatabaseVersion()).isEqualTo("n/a");
   }
 }

--- a/cloud-spanner-r2dbc/src/test/java/com/google/cloud/spanner/r2dbc/SpannerConnectionTest.java
+++ b/cloud-spanner-r2dbc/src/test/java/com/google/cloud/spanner/r2dbc/SpannerConnectionTest.java
@@ -400,6 +400,13 @@ public class SpannerConnectionTest {
     verify(this.mockClient).healthcheck(connection);
   }
 
+  @Test
+  public void getConnectionMetadata() {
+    SpannerConnection connection = new SpannerConnection(this.mockClient, TEST_SESSION, null);
+    assertThat(connection.getMetadata().getDatabaseProductName()).isEqualTo("Cloud Spanner");
+    assertThat(connection.getMetadata().getDatabaseVersion()).isEqualTo("n/a");
+  }
+
   private PartialResultSet makeBookPrs(String bookName) {
     return PartialResultSet.newBuilder()
         .setMetadata(ResultSetMetadata.newBuilder().setRowType(StructType.newBuilder()

